exporters/dynamicconfig/confignotifier.go

```diff
@@ -55,45 +68,114 @@ type ConfigWatcher interface {
 }
 
 type ConfigNotifier struct {
+	ch             chan struct{}
 	checkFrequency time.Duration
+	clock          controllerTime.Clock
 	config         *MetricConfig
+	configHost     string
+	lock           sync.Mutex
 	subscribed     map[ConfigWatcher]bool
+	ticker         controllerTime.Ticker
+	wg             sync.WaitGroup
 }
 
-// Constructor for a ConfigNotifier, also starts it running
-func New(checkFrequency time.Duration, config *MetricConfig) *ConfigNotifier {
+// Constructor for a ConfigNotifier
+// Set configHost to "" if there is no remote config service and the config is not dynamic
+func New(checkFrequency time.Duration, defaultConfig *MetricConfig, configHost string) *ConfigNotifier {
+	config := defaultConfig
+	if configHost != "" {
+		config = readConfig(configHost)
+	}
+
 	configNotifier := &ConfigNotifier{
+		ch:             make(chan struct{}),
 		checkFrequency: checkFrequency,
+		clock:          controllerTime.RealClock{},
 		config:         config,
+		configHost:     configHost,
 		subscribed:     make(map[ConfigWatcher]bool),
 	}
 
-	go configNotifier.checkChanges()
-
 	return configNotifier
 }
 
-func (notifier *ConfigNotifier) Register(watcher ConfigWatcher) {
-	notifier.subscribed[watcher] = true
-	watcher.OnInitialConfig(notifier.config)
+// SetClock supports setting a mock clock for testing.  This must be
+// called before Start().
+func (n *ConfigNotifier) SetClock(clock controllerTime.Clock) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.clock = clock
 }
 
-func (notifier *ConfigNotifier) Unregister(watcher ConfigWatcher) {
-	delete(notifier.subscribed, watcher)
+func (n *ConfigNotifier) Start() {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	if n.configHost == "" {
+		return
+	}
+
+	if n.ticker != nil {
+		return
+	}
+
+	n.ticker = n.clock.Ticker(n.checkFrequency)
+	n.wg.Add(1)
```

Comment:
Why is this needed? We allow a notifier to be started and stopped multiple times? If yes, why?

Reply:
I might have misunderstood your comment, so feel free to clarify! It's not necessary for a notifier to be able to stop and start.

The wg means that when we run ConfigNotifier.Stop(), it will wait for checkChanges to return before we Stop the entire thing. I added this because it seems more "correct" for the Stop method to stop everything associated with the ConfigNotifier, rather than letting a loose goroutine to continue.

If you mean why do we have a Stop() method at all, I decided to add it because it seems to be bad form for a goroutine to not have a stop condition, and it sounds plausible that at some point, we might want a stop mechanism. Let me know if you disagree!